### ISSUE TITLE
Child classes with concrete base classes do not serialize with correct Resource object

### ISSUE THREAD

**Issue by drewclauson** [steffenbrem/JsonApiBundle#18]
When a child class (bar) that inherits from a concrete base (foo) class is serialized, the JsonApiBundle resource annotation from the base class is loaded into the cache in place of the child class annotation and the base class annotation is never loaded into its correct class.

```
 /**
 * @JsonApi\Resource(type="foo")
 */
class foo
{
//...
}

 /**
 * @JsonApi\Resource(type="bar")
 */
class bar extend foo
{
//...
}
```

Result is that the "foo" will be loaded into the resource annotation for for the bar class instead of "bar".

I previously submitted PR #17 to resolve this in a basic manner for abstract classes, but I'm not sure how to handle the concrete class situation or if my original PR needs a different strategy.

Thanks!

**Comment by doug5791** [comment]
I have been looking at this issue for my current project and may have found a solution. 

In the file "MetadataFactory.php" I changed the first conditional of _getMetadataForClass($className)_ from :

```
if (isset($this->loadedMetadata[$className])) {         
            return $this->filterNullMetadata($this->loadedMetadata[$className]);
}
```

To:

```
if (isset($this->loadedClassMetadata[$className])) {        
            return $this->filterNullMetadata($this->loadedClassMetadata[$className]);
}
```

**Comment by drewclauson** [comment]
@doug5791 I don't have a good use case for a concrete base class at this point, but just wanted to mention that "MetadataFactory.php" is in the JMS Serializer bundle.  Would it work to extend that class in the JsonApiBundle and override the _getMetadataForClass($className)_ with your proposed change?  I have tried changing my code from abstract to concrete just to run some tests against this, but I ran into some undesirable serialization related to the fact that my child classes have a one to many relationship with each other (the first record in the relationship returned was listed as the base class type instead of the child class).  I should have more time to look at it later this week, though.  Let me know if you discover anything further.

**Comment by doug5791** [comment]
@drewclauson I do not see a reason you could not extend the class in JsonApiBundle. 

I will continue with the change I have made as to test if there are scenarios in which the change is not appropriate.

**Comment by doug5791** [comment]
@drewclauson I ran into an issue where when I had a collection of embedded models (i.e. User has many Roles) the first embedded model would serialize as having the parent classType; yet the rest of them would have the appropriate types. This is because the hierarchy of classes is being traversed from parent to child and escaping as soon as a particular condition is hit.

To remedy this issue  I commented out one more line in the function _getClassHierarchy($className)_

```
 private function getClassHierarchy($class)
    {
        $classes = array();
        $refl = new \ReflectionClass($class);

        do {
            $classes[] = $refl;
            $refl = $refl->getParentClass();
        } while (false !== $refl);

       // $classes = array_reverse($classes, false);

        if (!$this->includeInterfaces) {
            return $classes;
        }
    ...
```

which has the result of letting the hierarchy begin it's traversal from the child not the parent.

Note: These changes were made to the JMS package and may be extended in JsonApiBundle if you deem necessary. These are simply the fixes that work for me.
